cpp/doc/libcuspatial_refactoring_guide.md

```diff
@@ -0,0 +1,275 @@
+# cuSpatial C++ API Refactoring Guide
+
+The original cuSpatial C++ API (libcuspatial) was designed to depend on RAPIDS libcudf and use
+its core data types, especially `cudf::column`. For users who do not also use libcudf or other 
+RAPIDS APIS, depending on libcudf could be a big barrier to adoption of libcuspatial. libcudf is 
+a very large library and building it takes a lot of time.
+
+Therefore, we are developing a standalone libcuspatial C++ API that does not depend on libcudf. This 
+is a header-only template API with an iterator-based interface. This has a number of advantages
+
+  1. With a header-only API, users can include and build exactly what they use.
+  2. With a templated API, the API can be flexible to support a variety of basic data types, such 
+     as float and double for positional data, and different integer sizes for indices.
+  3. By templating on iterator types, cuSpatial algorithms can be fused with transformations of the
+     input data, by using "fancy" iterators. Examples include transform iterators and counting
+     iterators.
+  4. Memory resources only need to be part of APIs that allocate temporary intermediate storage.
+     Output storage need not be allocated inside the API using a memory resource, since
+
+The main disadvantages of this type of API are
+
+  1. Header-only APIs can increase compilation time for code that depends on them.
+  2. Some users (especially our Python API) may prefer a cuDF-based API.
+
+The good news is that by maintaining the existing libcudf-based C++ API as a layer above the header-
+only libcuspatial API, we can avoid problem 1 and problem 2 for users of the legacy API.
+
+## Example API
+
+Following is an example iterator-based API for `cuspatial::haversine_distance`.
```

Comment:
Totally tangential, but `cuspatial::` is right on the edge of being too long for a namespace imo. 

It's too bad [`cusp::`](https://github.com/cusplibrary/cusplibrary) is taken by a sparse linear algebra library that also happens to build on Thrust.